README.md
# Stable Cache

[![npm](https://img.shields.io/npm/v/stable-cache.svg)](https://www.npmjs.com/package/stable-cache)
[![Dependency Status](https://david-dm.org/joao-fontenele/stable-cache.svg)](https://david-dm.org/joao-fontenele/stable-cache)
[![devDependency Status](https://david-dm.org/joao-fontenele/stable-cache/dev-status.svg)](https://david-dm.org/joao-fontenele/stable-cache#info=devDependencies)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![Test Coverage](https://api.codeclimate.com/v1/badges/4a36da1b2ed3bc53f940/test_coverage)](https://codeclimate.com/github/joao-fontenele/stable-cache/test_coverage)
[![Maintainability](https://api.codeclimate.com/v1/badges/4a36da1b2ed3bc53f940/maintainability)](https://codeclimate.com/github/joao-fontenele/stable-cache/maintainability)
[![Gitter chat](https://badges.gitter.im/gitterHQ/gitter.png)](https://gitter.im/stable-cache/community)

This lib makes extensive use of a concept called producer. A producer is the source of truth of a cached value. Producers are only called when needed to set a cache, and are usually not needed when there's a cache hit.

The proposal of this library is to make it easier to manage producers and how keys are fetched/saved to redis.

Features:

- there's easily configurable resilience against producer failures:
  - timeout
  - retry with exponential backoff, and [decorrelated jitter](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/)
  - circuit breaker
- optionally set redis ttl for produced values
- optionally ignoring cached values and forcing values refresh
- optionally refreshing keys in background. Useful when you want keys to be updated before their ttl would evict them.
- optionally producing values in background, while returning the cached value, even on cache misses
- builtin RTA for prometheus

## Usage

### Cache instance

If the `circuitBreaker` config is not provided this cache instance simply will not have a circuit breaker flow, and will still be usable.

I recommend to instantiate a cache for each service you consume, or else the circuit breaker would be shared. Which in turn could mean that a failing service would open the circuit for working services too.

```js
const Redis = require('ioredis');
const { Cache } = require('stable-cache');

const redis = new Redis({
  port: 6379,
  host: 'redis', // localhost
});

const cache = new Cache({
  redis, // required to inject a redis client
  options: { // everything is optional from here
    name: 'someService', // name of the service the cache will handle, useful for RTA
    circuitBreaker: { // circuit breaker config for the service
      // percentage of failing producer requests to trigger the circuit breaker
      threshold: 0.2,
      // time window to be considered by the threshold, in milliseconds
      duration: 10000,
      // attempt to half open the circuit after this time in milliseconds
      halfOpenAfter: 20000,
      // don't open the circuit if there's less than this amount of RPS.
      // This avoids opening the circuit, during failures in low load periods.
      minimumRps: 5,
    },
  },
});
```

### Cache.get method

Only the `key` argument is required. Other options are used to augment the producer flow. See more usage examples at `lib/examples/cache-get-usage.js`

Cache get options are:

> NOTE: circuit breaker policy is configured only when instantiating the `Cache` class.

| Option | Description | Default Behavior | Default Value |
| -: | :- | :- | :- |
| `producer` | A function that will be called when the source of the truth of the key is needed, has the signature `function(): Promise<string>` | No producer to call | `null` |
| `ttl` | Time in milliseconds for the redis key ttl, in case the producer resolves | No ttl set | `null` |
| `producerRetry` | Object configuration for the exponential backoff | No retry | `null` |
| `producerRetry.maxDelay` | Max amount of delay between retry attempts, in milliseconds | - | `30000` |
| `producerRetry.maxAttempts` | Max amount of retries for the producer | - | `10` |
| `producerRetry.initialDelay` | Initial delay, in milliseconds before the retry | - | `128` |
| `producerTimeout` | Timeout in milliseconds for the producer. Even if there's a timeout, if the producer eventually resolves, the key will be set in background so it would be best if you configure a greater timeout on the producer itself, together with this config | No timeout | `null` |
| `returnEarlyFromCache` | Whether to return the cached value, and make the producer call on background, on cache miss | Await the producer on cache miss | `false` |
| `overrideCache` | Whether to ignore cached values and request producer, regardless of cache hits | Don't ignore cache hits, and call producer only if needed | `false` |
| `shouldRefreshKey` | A callback that if returns `true`, calls the producer and sets the key on background. Has the signature `function(key, currentTTL, options): boolean` | No automatic refresh of keys | `null` |

Example of all the configs:

```js
function producer() {
  return new Promise((resolve) => {
    setTimeout(() => resolve('some-value'), 2000);
  });
}

cache.get(
  'some^key', // required redis key to get
  { // optional configs
    producer,
    ttl: 1000,
    producerRetry: {
      maxDelay: 30000,
      maxAttempts: 10,
      initialDelay: 128
    },
    producerTimeout: 1000,
    returnEarlyFromCache: false,
    overrideCache: false,
    shouldRefreshKey(key, currentTTL, options) {
      if (!options.ttl || currentTTL <= 0) {
        return false;
      }
      // options.ttl is the usually configured ttl for the given key
      const halfLife = options.ttl / 2;
      // refresh key if currentTTL is 50% below the initial ttl
      return currentTTL <= halfLife;
    }
  },
);
```

### Cache.set method

Sets a key with a value, and receives an optional ttl for the key.


Cache set options:

| Option | Description | Default Behavior | Default Value |
| -: | :- | :- | :- |
| `ttl` | Time in milliseconds for the redis key ttl | No ttl set | `null` |

Example:

```js
// sets `some^key` with value `a-value` with ttl of 30 seconds
cache.set(
  'some^key', // required key
  'a-value', // required value
  { // optional configs
    ttl: 30000, // 30 seconds
  }
);
```

## Prometheus Exporter

There's an exporter for prometheus that exposes the following metrics

- `cache_results_count`: counts cache hits/misses, with labels: `['service', 'operation', 'result']`
- `cache_operations_duration_seconds`: histogram for cache RT, with labels: `['service', 'operation']`
- `cache_operations_count`: counts cache operations made, with labels: `['service', 'operation']`
- `producer_operations_duration_seconds`: histogram with RT for producer calls, with labels: `['service']`
- `producer_operations_result_count`: counts the successes/errors for producer calls, with labels: `['service', 'result']`
- `producer_circuit_break_state`: gauge for the state of a service circuit breaker. A value of `0` means the circuit is closed (working normally), and a value of `1` means the circuit is open (fail fast is activated).

What labels mean?

- `service`: is the cache `name` option
- cache `operation`: type of redis operation, e.g. `set`, `get`.
- cache `result`: cache `hit` or `miss`
- producer `result`: `success`, or `error`

### Usage

Prometheus exporter options for the constructor are:

| Option | Description | Default Behavior | Default Value |
| -: | :- | :- | :- |
| `prefix` | Prefix to be added to every exported metric | No prefix added | `''` |
| `registers` | Array of prometheus registers to which metrics should be exported | Use default `prom-client` register | `[Prometheus.register]` |
| `cacheBuckets` | Array of numbers, representing the histogram buckets for cache RT | Use default value | `Prometheus.exponentialBuckets(0.05, 2, 8)` |
| `producerBuckets` | Array of numbers, representing the histogram buckets for producer RT | Use default value | `Prometheus.exponentialBuckets(0.1, 2, 8)` |

Example:

```js
const Prometheus = require('prom-client');
const { PrometheusExporter } = require('stable-cache');

const exporter = new PrometheusExporter({
  prefix: 'my_app_',
  registers: [Prometheus.register],
  cacheBuckets: Prometheus.exponentialBuckets(0.05, 2, 8),
  producerBuckets: Prometheus.exponentialBuckets(0.1, 2, 8),
});

exporter.collectMetrics(); // start collecting metrics
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.1](https://github.com/joao-fontenele/stable-cache/compare/v0.1.0...v0.1.1) (2020-10-22)

## [0.1.0](https://github.com/joao-fontenele/stable-cache/compare/v0.0.1...v0.1.0) (2020-06-01)


### ⚠ BREAKING CHANGES

* exporting classes Cache, PrometheusExporter, Producer, and
  RTAEmitter as properties of their modules

### Features

* make this a typescript project ([bea5526](https://github.com/joao-fontenele/stable-cache/commit/bea55262dfc37f022265e513f592af896fb429fa))

### [0.0.1](https://github.com/joao-fontenele/stable-cache/compare/v0.0.0...v0.0.1) (2020-05-18)

## 0.0.0 (2020-05-17)


### Features

* add cockatiel lib ([f7f5fae](https://github.com/joao-fontenele/stable-cache/commit/f7f5fae6d8d3b7aa16226b31725eee79c3118bde))
* **producer:** add rta for circuit breaker state changes ([adcc2ae](https://github.com/joao-fontenele/stable-cache/commit/adcc2ae2768b5789bf0c352a76566245628ac947))
* initial commit ([975b670](https://github.com/joao-fontenele/stable-cache/commit/975b670461c813f5ebd6bc1313488bbcc09a07ba))
* **cache:** add methods get and set ([33f3379](https://github.com/joao-fontenele/stable-cache/commit/33f33799241108579da6d41c39a2c7f6a7c85417))
* **cache:** optional config to refresh a key ([bd308d2](https://github.com/joao-fontenele/stable-cache/commit/bd308d22e3024e7dbfe630d21f4edfbb9d96e2b9)), closes [#9](https://github.com/joao-fontenele/stable-cache/issues/9)
* **cache:** use rtaEmitter in cache methods ([9f950c0](https://github.com/joao-fontenele/stable-cache/commit/9f950c0b8cd52845ca9e3e5dcacc9076b249861f))
* **classes/cache:** use producer class ([cf33733](https://github.com/joao-fontenele/stable-cache/commit/cf3373369eb522660a8873b3a15b5a4bc1757936))
* install prom-client as a peer dependency ([fc49b36](https://github.com/joao-fontenele/stable-cache/commit/fc49b36a5f50cc5ad03cc9ed1e82618c36eabd61))
* **classes/producer:** add producer resilience class ([b995c01](https://github.com/joao-fontenele/stable-cache/commit/b995c018cf99f02ace68c011bb5b2b6e9af770f9))
* **client/redis:** local redis client ([045db7c](https://github.com/joao-fontenele/stable-cache/commit/045db7ca7db623964de1f46bb7c530127a20db28))
* project's main file ([4c6efe3](https://github.com/joao-fontenele/stable-cache/commit/4c6efe379db0a61daa7e5048f6603f300be9d576))
* **examples:** formalize an examples folder ([243d6f2](https://github.com/joao-fontenele/stable-cache/commit/243d6f2306c43a3f0fadcb9751bf9f1cbfd5425d))
* **policies:** add separated policy classes ([38e7b9a](https://github.com/joao-fontenele/stable-cache/commit/38e7b9a5a0f237b69346a80b9c20f94d6488ed0c))
* **prometheus-exporter:** class to export metrics for prometheus ([3f852e5](https://github.com/joao-fontenele/stable-cache/commit/3f852e577f70eea05c9c22de13f67964f7f49485))
* **rta-emitter:** add class to help rta cache/producer operations ([c87d681](https://github.com/joao-fontenele/stable-cache/commit/c87d681bdfc9c8273483313c36d2bf32286cc86d))


### Bug Fixes

* **cache:** pass missing service name to producer class ([58ae330](https://github.com/joao-fontenele/stable-cache/commit/58ae3304fedcc4f688af921fcdae8f0c6f4ad62d))
* **ci:** add whole lib folder in coverage reports ([3c6a4b2](https://github.com/joao-fontenele/stable-cache/commit/3c6a4b28df4df6e58b0eb6eb66dd5d9839499d3a))
* change lots of files so CB state change events, work ([1f4a8f6](https://github.com/joao-fontenele/stable-cache/commit/1f4a8f6a982d36e0c1e2180f8554b2f352c62505)), closes [#14](https://github.com/joao-fontenele/stable-cache/issues/14)


examples/server.js
const Bluebird = require('bluebird');
const express = require('express');
const Prometheus = require('prom-client');

const { Cache, PrometheusExporter } = require('../dist');
const Redis = require('./redis');

const app = express();

const PORT = 3000;

let cache;
const exporter = new PrometheusExporter();
exporter.collectMetrics();

async function success(time, value) {
  return Bluebird.delay(time || 1)
    .then(() => value);
}

async function fail(time) {
  return Bluebird.delay(time || 1)
    .then(() => {
      throw new Error('failed!');
    });
}

function getRandomInBetween(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

app.get('/success', async (req, res) => {
  console.log('GET /success');
  const producer = () => success(getRandomInBetween(150, 750), 'some key value');
  try {
    const value = await cache.get(
      'service',
      {
        producer,
        ttl: 1000,
        producerRetry: { maxDelay: 5000, maxAttempts: 5 },
        producerTimeout: 1000,
      },
    );
    res.json({ value });
  } catch (err) {
    console.log('err', err);
    res.status(500).json({ err });
  }
});

app.get('/failure', async (req, res) => {
  console.log('GET /failure');
  const producer = () => fail(1);
  try {
    const value = await cache.get(
      'service',
      {
        producer,
        ttl: 1000,
        producerRetry: { maxDelay: 5000, maxAttempts: 5 },
        producerTimeout: 1000,
      },
    );
    res.json({ value });
  } catch (err) {
    res.status(500).json({ err: err.message });
  }
});

app.get('/metrics', (req, res) => {
  console.log('GET /metrics');
  res.set('Content-Type', Prometheus.register.contentType);
  return res.end(Prometheus.register.metrics());
});


Redis.connect()
  .then(() => {
    cache = new Cache({
      redis: Redis.redis,
      options: {
        name: 'testService',
        circuitBreaker: {
          threshold: 0.2,
          duration: 10000,
          halfOpenAfter: 20000,
        },
      },
    });
    app.listen(PORT, () => {
      console.log(`listening on port ${PORT}`);
    });
  });
